contracts/psp22/lib.rs

```diff
@@ -0,0 +1,153 @@
+        spender: AccountId,
+        value: Balance,
+    }
+
+    #[ink(event)]
+    pub struct Transfer {
+        #[ink(topic)]
+        from: Option<AccountId>,
+        #[ink(topic)]
+        to: Option<AccountId>,
+        value: Balance,
+    }
+
+    #[ink(storage)]
+    #[derive(Default)]
+    pub struct Token {
+        total_supply: Balance,
+        balances: Mapping<AccountId, Balance>,
+        allowances: Mapping<(AccountId, AccountId), Balance>,
+    }
+
+    pub type Event = <Token as ContractEventBase>::Type;
+
+    impl Token {
+        #[ink(constructor)]
+        pub fn new(total_supply: Balance) -> Self {
+            let mut balances = Mapping::default();
+            let caller = Self::env().caller();
+            balances.insert(caller, &total_supply);
+            Self {
+                total_supply,
+                balances,
+                allowances: Default::default(),
+            }
+        }
+
+        fn transfer_from_to(
+            &mut self,
+            from: &AccountId,
+            to: &AccountId,
+            value: Balance,
+        ) -> Result<(), PSP22Error> {
+            let from_balance = self.balance_of(*from);
+            if from_balance < value {
+                return Err(PSP22Error::InsufficientBalance);
+            }
+
+            self.balances.insert(from, &(from_balance - value));
+            let to_balance = self.balance_of(*to);
+            self.balances.insert(to, &(to_balance + value));
+
+            Self::emit_event(
+                self.env(),
+                Event::Transfer(Transfer {
+                    from: Some(*from),
+                    to: Some(*to),
+                    value,
+                }),
+            );
+
+            Ok(())
+        }
+
+        fn emit_event<EE>(emitter: EE, event: Event)
+        where
+            EE: EmitEvent<Self>,
+        {
+            emitter.emit_event(event);
+        }
+    }
+
+    impl PSP22 for Token {
+        /// Returns the total token supply.
+        #[ink(message)]
+        fn total_supply(&self) -> Balance {
+            self.total_supply
+        }
+
+        /// Returns the account balance for the specified `owner`.
+        #[ink(message)]
+        fn balance_of(&self, owner: AccountId) -> Balance {
+            self.balances.get(owner).unwrap_or_default()
+        }
+
+        /// Returns the amount which `spender` is allowed to withdraw on behalf of the `owner` account.
+        #[ink(message)]
+        fn allowance(&self, owner: AccountId, spender: AccountId) -> Balance {
+            self.allowances.get((owner, spender)).unwrap_or_default()
+        }
+
+        /// Allows `spender` to withdraw from the caller's account multiple times, up to the `value` amount.
+        #[ink(message)]
+        fn approve(&mut self, spender: AccountId, value: Balance) -> Result<(), PSP22Error> {
+            let owner = self.env().caller();
+            self.allowances.insert((&owner, &spender), &value);
```

Comment:
not sure whether we should check if owner has at least `value`... I see why it might be useful to not check it, but wondering about security of it

Reply:
will be checked at transfer time, that's all that matters. I can think of plenty of use-cases that'd be blocked if you check it here - certain types of flashloans for example